src/infrastructure/arranging/sequencing.rs

```diff
@@ -0,0 +1,236 @@
+    fn stream_count(&self) -> usize {
+        self.streams.len()
+    }
+
+    /// Try to get an [`SequencingStream`](./struct.SequencingStream.html) by `stream_id`.
+    /// When the stream does not exist, it will be inserted by the given `stream_id` and returned.
+    fn get_or_create_stream(&mut self, stream_id: u8) -> &mut Self::Stream {
+        self.streams
+            .entry(stream_id)
+            .or_insert_with(|| SequencingStream::new(stream_id))
+    }
+}
+
+/// A stream on which items will be arranged in-sequence.
+///
+/// # Algorithm
+///
+/// With every sequencing operation an `top_index` is given.
+///
+/// There are two scenarios that are important to us.
+/// 1. `incoming_index` >= `top_index`.
+/// This item is the newest or newer than the last one we have seen.
+/// Because of that we should return it back to the user.
+/// 2. `incoming_index` < `top_index`.
+/// This item is older than the newest item we have seen so far.
+/// Since we don't care about old items we can toss it a way.
+///
+/// # Remarks
+/// - See [super-module](../index.html) for more information about streams.
+pub struct SequencingStream<T> {
+    // the id of this stream.
+    stream_id: u8,
+    // the highest seen item index.
+    top_index: usize,
+    // I need `PhantomData`, otherwise, I can't use a generic in the `Arranging` implementation because `T` is not constrained.
+    phantom: PhantomData<T>
+}
+
+impl<T> SequencingStream<T> {
+    /// Constructs a new, empty '[SequencingStream](./struct.SequencingStream.html)'.
+    ///
+    /// The default stream will have a capacity of 32 items.
+    pub fn new(stream_id: u8) -> SequencingStream<T> {
+        SequencingStream {
+            stream_id,
+            top_index: 0,
+            phantom: PhantomData
+        }
+    }
+
+    /// Returns the identifier of this stream.
+    fn stream_id(&self) -> u8 {
+        self.stream_id
+    }
+}
+
+impl<T> Arranging for SequencingStream<T> {
+    type ArrangingItem = T;
+
+    /// Will arrange the given item based on a sequencing algorithm.
+    ///
+    /// With every sequencing operation an `top_index` is given.
+    ///
+    /// # Algorithm
+    ///
+    /// There are two scenarios that are important to us.
+    /// 1. `incoming_index` >= `top_index`.
+    /// This item is the newest or newer than the last one we have seen.
+    /// Because of that we should return it back to the user.
+    /// 2. `incoming_index` < `top_index`.
+    /// This item is older than we the newest packet we have seen so far.
+    /// Since we don't care about old items we can toss it a way.
+    ///
+    /// # Remark
+    /// - All old packets will be tossed away.
+    /// - None is returned when an old packet is received.
+    fn arrange(
+        &mut self,
+        incoming_index: usize,
+        item: Self::ArrangingItem,
+    ) -> Option<Self::ArrangingItem> {
+        if incoming_index >= self.top_index {
```

Comment:
You likely just want `>` here.

Reply:
Good one, I initially when with the expected sequence. That is why I had >= there.